### PR TITLE
Integration test for IOT Hub manager

### PR DESCRIPTION
**Pull request by sushilraje**
# Description and Motivation 

...

# Change type 

- [ ] Bug fix
- [x] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

**Checklist:**

- [ ] All tests passed
- [ ] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: IoTHubManager/Constants.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
+﻿using System;
```
Comment (sushilraje) [inline IoTHubManager/Constants.cs]: Currently, I am having this file separate from others. Will merge with the file under Helpers proj, once it is available.

File: IoTHubManager/CreateDeviceTests.cs [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+namespace IoTHubManager
+{
+    [Collection("IoTHub Manager Tests")]
+    public class CreateDeviceTest
+    {
+
+        internal HttpRequestWrapper Request;
+
+        private string DEVICE_TEMPLATE_AUTO_GEN_AUTH,
```
Comment (ppathan) [inline IoTHubManager/CreateDeviceTests.cs]: nit: declare string at each line

File: IoTHubManager/CreateDeviceTests.cs [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        /// Tests for creation of devices with all permutation of AUTH and Id creation
+        /// </summary>
+        
+
+        /**
+         * Creates a device with auto generated id and symmetric auth.
+         * Verifies that the resultant device has system-generated
+         * id as well as auth credentails.
+         */
+        [Fact, Trait(Constants.TEST , Constants.INTEGRATION_TEST )]
```
Comment (ppathan) [inline IoTHubManager/CreateDeviceTests.cs]: nit: Formatting seems wrong for all the attributes no space before `,` and not space before `)`. Here all the other places

File: IoTHubManager/CreateDeviceTests.cs [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        public void Creates_Device_with_AutoGenerated_Id_and_Auth()
```
Comment (ppathan) [inline IoTHubManager/CreateDeviceTests.cs]: nit: Is this this the naming convention that we finalized on?

File: IoTHubManager/CreateDeviceTests.cs [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+            // Assert device ID and auth or not null OR empty and
+            // other required properties are set. Also check auth type.
+            Assert.False(string.IsNullOrEmpty(createdDeviceId));
+            Assert.False(createdDevice["IsSimulated"].ToObject<bool>());
+            Assert.True(createdDevice["Enabled"].ToObject<bool>());
+            Assert.Equal(Constants.Auth.SYMMETRIC, authentication["AuthenticationType"]);
+            Assert.False(string.IsNullOrEmpty(primaryKey));
+            Assert.False(string.IsNullOrEmpty(secondaryKey));
+
```
Comment (ppathan) [inline IoTHubManager/CreateDeviceTests.cs]: nit: Extra space

File: IoTHubManager/CreateDeviceTests.cs [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        /**
+         * Creates a device with custom id and symmetric auth
+         * credentails. Verifies that the resultant device has 
+         * correct id and auth credentails.
+         */
+        [Fact, Trait(Constants.TEST , Constants.INTEGRATION_TEST )]
+        public void Creates_Device_with_Custom_Id_and_Auth()
+        {
+
```
Comment (ppathan) [inline IoTHubManager/CreateDeviceTests.cs]: nit: extra space

File: IoTHubManager/CreateDeviceTests.cs [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+            // Assert device ID and auth or not null OR empty and
+            // other required properties are set. Also check auth type.
+            Assert.Equal(createdDeviceId, id);
+            Assert.Equal(Constants.Auth.SYMMETRIC, authentication["AuthenticationType"]);
+            Assert.Equal(primaryKey, createdPrimaryKey);
+            Assert.Equal(secondaryKey, createdSecondaryKey);
+            Assert.False(createdDevice["IsSimulated"].ToObject<bool>());
+            Assert.True(createdDevice["Enabled"].ToObject<bool>());
+
```
Comment (ppathan) [inline IoTHubManager/CreateDeviceTests.cs]: nit: extra line

File: IoTHubManager/CreateDeviceTests.cs [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+            Assert.False(string.IsNullOrEmpty(createdDeviceId));
+            Assert.Equal(Constants.Auth.X509, authentication["AuthenticationType"]);
+            Assert.Equal(primaryThumbprint, createdPrimaryThumbprint);
+            Assert.Equal(secondaryThumbprint, createdSecondaryThumbprint);
+            Assert.False(createdDevice["IsSimulated"].ToObject<bool>());
+            Assert.True(createdDevice["Enabled"].ToObject<bool>());
+
+        }
+
+
```
Comment (ppathan) [inline IoTHubManager/CreateDeviceTests.cs]: nit: extra lines

File: IoTHubManager/CreateDeviceTests.cs [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+            // Assert device ID and auth or not null OR empty and
+            // other required properties are set. Also check auth type.
+            Assert.False(string.IsNullOrEmpty(createdDeviceId));
+            Assert.Equal(Constants.Auth.X509, authentication["AuthenticationType"]);
+            Assert.Equal(primaryThumbprint, createdPrimaryThumbprint);
+            Assert.Equal(secondaryThumbprint, createdSecondaryThumbprint);
+            Assert.False(createdDevice["IsSimulated"].ToObject<bool>());
+            Assert.True(createdDevice["Enabled"].ToObject<bool>());
+
```
Comment (ppathan) [inline IoTHubManager/CreateDeviceTests.cs]: not : extra line

File: IoTHubManager/CreateJobsTest.cs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+            Assert.Equal(HttpStatusCode.OK, tagJobResponse.StatusCode);
+
+            var tagJob = JObject.Parse(tagJobResponse.Content);
+
+            Assert.Equal<int>(Constants.Jobs.JOB_IN_PROGRESS, tagJob["Status"].ToObject<int>());
+            Assert.Equal<int>(Constants.Jobs.TAG_JOB, tagJob["Type"].ToObject<int>());
+
+            var tagJobStatus = Helpers.ReTry_GetJobStatus(Request, tagJob["JobId"].ToString());
+
+            Assert.Equal<int>(Constants.Jobs.JOB_COMPLETED, tagJobStatus["Status"].ToObject<int>());   //Assert to see if Last try yielded correct status.
```
Comment (ppathan) [inline IoTHubManager/CreateJobsTest.cs]: nit: move the comment to previous line

File: IoTHubManager/Helpers.cs [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            IHttpResponse jobStatusResponse = Request.Get(JobId, null);
+            Assert.Equal(HttpStatusCode.OK, jobStatusResponse.StatusCode);
+            return JObject.Parse(jobStatusResponse.Content);
+        }
+
+        /**
+         * Monitor job status using polling (re-try) mechanism 
+         */
+        internal static JObject ReTry_GetJobStatus(HttpRequestWrapper Request, string jobId)
```
Comment (ppathan) [inline IoTHubManager/Helpers.cs]: nit: `Retry` is one word and should be `GetJobStatusWithRetry`

File: IoTHubManager/run.sh [hunk 12]
```diff
@@ -17,7 +17,7 @@ start_containers() {
     ./scripts/storageadapter.sh start
     ./scripts/iothubmanager.sh start
     ./scripts/devicesimulation.sh start
-    ./scripts/pcsconfig.sh start
+    #./scripts/pcsconfig.sh start
```
Comment (ppathan) [inline IoTHubManager/run.sh]: If we don't need the pcsconfig.sh then remove the line

File: IoTHubManager/run.sh [hunk 13]
```diff
@@ -29,7 +29,7 @@ stop_containers() {
     ./scripts/storageadapter.sh stop
     ./scripts/iothubmanager.sh stop
     ./scripts/devicesimulation.sh stop
-    ./scripts/pcsconfig.sh stop
+    #./scripts/pcsconfig.sh stop
```
Comment (ppathan) [inline IoTHubManager/run.sh]: remove

File: IoTHubManager/Constants.cs [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace IoTHubManager
+{
+    class Constants
```
Comment (molly-moen) [inline IoTHubManager/Constants.cs]: we have a constants file in Helpers, move this to that file

File: IoTHubManager/Constants.cs [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+                DEVICE_FILE_X509_AUTH = DEVICES_DIR + "Device_Template_X509_Auth.json";
+
+            //Job Files (JSON object holding job (tag/reconfigure) descriptions.)
+            public const string
+                TAGS_FILE = JOBS_DIR + "Tags.json",
+                REBOOT_METHOD_FILE = JOBS_DIR + "Method_Chiller_Reboot.json",
+                FIRMWAREUPDATE_METHOD_FILE = JOBS_DIR + "Method_Elevator_FrimwareUpdate.json",
+                RECONFIGURE_DEVICE_FILE = JOBS_DIR + "Reconfigure_Chiller_Update_Model.json",
+                SIMULATION_FILE = JOBS_DIR + "Simulation.json";
+
```
Comment (molly-moen) [inline IoTHubManager/Constants.cs]: nit: remove unnecessary blank lines (here and where you have 2 blank lines in a row in this file)

File: IoTHubManager/CreateDeviceTests.cs [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+            //Create request wrapper object for interacting with iothub manager microservices for CRUD on devices
+            this.Request = new HttpRequestWrapper(Constants.Urls.IOTHUB_ADDRESS, Constants.Urls.DEVICE_PATH);
+
+            //Fetch different device templates
+            DEVICE_TEMPLATE_AUTO_GEN_AUTH = System.IO.File.ReadAllText(Constants.Path.DEVICE_FILE_AUTO_GEN_AUTH);
+            DEVICE_TEMPLATE_SYMMETRIC_AUTH = System.IO.File.ReadAllText(Constants.Path.DEVICE_FILE_SYMMETRIC_AUTH);
+            DEVICE_TEMPLATE_X509_AUTH = System.IO.File.ReadAllText(Constants.Path.DEVICE_FILE_X509_AUTH);                     
+        }
+
+        /// <summary>
```
Comment (molly-moen) [inline IoTHubManager/CreateDeviceTests.cs]: should this comment be here?

File: IoTHubManager/CreateDeviceTests.cs [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+            Assert.True(createdDevice["Enabled"].ToObject<bool>());
+        }
+
+        /**
+         * Creates a device with custom id and X509 auth credentails.
+         * Verifies that the resultant device has correct id
+         * and auth credentails.
+         */
+        [Fact, Trait(Constants.TEST, Constants.INTEGRATION_TEST)]
+        public void DeviceCreated_IfCustomIdandX509AuthPassed()
```
Comment (molly-moen) [inline IoTHubManager/CreateDeviceTests.cs]: nit: capitalize and

File: IoTHubManager/CreateDeviceTests.cs [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+                                                     .Replace(Constants.Keys.PRIMARY_TH, primaryThumbprint)
+                                                     .Replace(Constants.Keys.SECONDARY_TH, secondaryThumbprint);
+
+            // Act
+            var response = Request.Post(device);
+
+            // Asserts 
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+
+            var createdDevice = JObject.Parse(response.Content);
```
Comment (molly-moen) [inline IoTHubManager/CreateDeviceTests.cs]: can you extract some of these checks into a private method? It looks like they are pretty similar

File: IoTHubManager/Constants.cs (hunk 1 above)
Comment (molly-moen) [inline IoTHubManager/Constants.cs]: add // Copyright (c) Microsoft. All rights reserved.

Comment (molly-moen) [reply IoTHubManager/Constants.cs]: also run remove and sort usings on this and all other files

File: IoTHubManager/CreateJobsTest.cs [hunk 19]
```diff
@@ -0,0 +1,124 @@
+﻿using System;
```
Comment (molly-moen) [inline IoTHubManager/CreateJobsTest.cs]: // Copyright (c) Microsoft. All rights reserved.

File: IoTHubManager/CreateJobsTest.cs [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+        public CreateJobsTest()
+        {
+            this.Request = new HttpRequestWrapper(Constants.Urls.IOTHUB_ADDRESS, Constants.Urls.JOBS_PATH);
+
+            Simulation simulation = Simulation.GetSimulation();
+
+            simulatedDeviceId = Constants.SimulatedDevices.SIMULATED_DEVICE + "." + simulation.healthyDeviceNo.ToString();
+            simulatedFaultyDeviceId = Constants.SimulatedDevices.SIMULATED_FAULTY_DEVICE + "." + simulation.faultyDeviceNo.ToString();
+        }
+
```
Comment (molly-moen) [inline IoTHubManager/CreateJobsTest.cs]: remove extra blank lines (only 1 between methods)

File: IoTHubManager/Helpers.cs [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+namespace IoTHubManager
+{
+    class Helpers
+    {
+        private readonly HttpRequestWrapper Request;
+
+        /*
+        Generates random SHA1 hash mimicing the X509 thumb print
+         */
+        internal static string GenerateNewThumbPrint()
```
Comment (molly-moen) [inline IoTHubManager/Helpers.cs]: nit: Thumbprint

File: IoTHubManager/Helpers.cs [hunk 22]
```diff
@@ -0,0 +1,85 @@
+﻿using System;
```
Comment (molly-moen) [inline IoTHubManager/Helpers.cs]: // Copyright (c) Microsoft. All rights reserved.
remove and sort usings

File: IoTHubManager/HttpRequestWrapper.cs [hunk 23]
```diff
@@ -0,0 +1,65 @@
+﻿using System;
```
Comment (molly-moen) [inline IoTHubManager/HttpRequestWrapper.cs]: should this go into generic helper project?

File: scripts/build [hunk 24]
```diff
@@ -14,12 +14,12 @@ fi
 
 echo "The repo name is: " $2
 
+./IoTHubManager/run.sh "IoT Hub Manager tests"
```
Comment (molly-moen) [inline scripts/build]: why was this moved?

Comment (sushilraje) [reply scripts/build]: Nothing much. I wanted to test my build failure.

File: IoTHubManager/HttpRequestWrapper.cs (hunk 23 above)
Comment (sushilraje) [reply IoTHubManager/HttpRequestWrapper.cs]: I m not sure if everyone needs this.

File: IoTHubManager/CreateDeviceTests.cs (hunk 18 above)
Comment (sushilraje) [reply IoTHubManager/CreateDeviceTests.cs]: Yes, I did those. Now they are Helpers.Device

File: IoTHubManager/HttpRequestWrapper.cs (hunk 23 above)
Comment (molly-moen) [reply IoTHubManager/HttpRequestWrapper.cs]: It's generic enough that I would move it

Comment (sushilraje) [reply IoTHubManager/HttpRequestWrapper.cs]: Ok

File: IoTHubManager/Constants.cs [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+            public const string PropertyValue = "{PropertyValue}";
+        }
+
+        public class Auth
+        {
+            public const int SYMMETRIC = 0;
+            public const int X509 = 1;
+        }
+
+        //Simulated devices on which Job test(s) are run on.
```
Comment (ppathan) [inline IoTHubManager/Constants.cs]: nit: space after comment

File: IoTHubManager/Constants.cs [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+        public class SimulatedDevices
+        {
+            public const string SIMULATED_DEVICE = "chiller-01";            //Non faulty
+            public const string SIMULATED_FAULTY_DEVICE = "elevator-02";    //Faulty
```
Comment (ppathan) [inline IoTHubManager/Constants.cs]: nit: ditto about space after comment

File: IoTHubManager/CreateDeviceTests.cs [hunk 27]
```diff
@@ -0,0 +1,194 @@
...
+            // Asserts
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            var createdDevice = JObject.Parse(response.Content);
+            deviceId = createdDevice["Id"].ToString();
+            Helper.Device.AssertCommonDeviceProperties(null, createdDevice);
+            Helper.Device.AssertX509Authentication(primaryThumbprint, secondaryThumbprint, createdDevice);
+            Helper.Device.CheckIfDeviceExists(this.Request, createdDevice);
+        }
+
+        public void Dispose()
```
Comment (ppathan) [inline IoTHubManager/CreateDeviceTests.cs]: Is Dispose called on destruction?

File: IoTHubManager/DevicePropertiesTest.cs [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+            var response = Request.Put(this.deviceId, this.device);
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            Thread.Sleep(Constants.Jobs.WAIT*5);
+
+            // Asserts
+            AssertKeyExistsInPropertyList(devicePropertyKey);
+            // DeleteDeviceProperty(devicePropertyKey); Deleting tags NOT working
+            // Thread.Sleep(Constants.Jobs.WAIT * 5);
+            // AssertKeyDoesNotExistsInPropertyList(devicePropertyKey);
+            // Deleting tags si currently not updating the device property list.
```
Comment (ppathan) [inline IoTHubManager/DevicePropertiesTest.cs]: typo: is instead of si.
`There is a bug in delete tags where it appends new tags but doesn't remove deleted tags. There is a bug open for it`

File: IoTHubManager/DevicePropertiesTest.cs [hunk 29]
```diff
@@ -0,0 +1,117 @@
...
+            //Arrange
+            string jobId = Guid.NewGuid().ToString();
+            string devicePropertyKey = Guid.NewGuid().ToString();
+            string devicePropertyValue = Guid.NewGuid().ToString();
+            this.device["Tags"][devicePropertyKey] = devicePropertyValue;
+
+            // Act
+            var response = Request.Put(this.deviceId, this.device);
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            Thread.Sleep(Constants.Jobs.WAIT*5);
```
Comment (ppathan) [inline IoTHubManager/DevicePropertiesTest.cs]: nit: space before and after *. Also if multiplying by 5 then should WAIT already have that value?

File: IoTHubManager/resources/jobs/Simulation.json [hunk 30]
```diff
@@ -0,0 +1,13 @@
+﻿{
```
Comment (ppathan) [inline IoTHubManager/resources/jobs/Simulation.json]: nit: tab size is wrong in this file. Should be 4

File: IoTHubManager/resources/jobs/Tags.json [hunk 31]
```diff
@@ -0,0 +1,14 @@
+﻿{
```
Comment (ppathan) [inline IoTHubManager/resources/jobs/Tags.json]: tabs

File: scripts/build [hunk 32]
```diff
@@ -14,12 +14,13 @@ fi
 
 echo "The repo name is: " $2
 
-./StorageAdapter/run.sh "Storage Adapter tests"
-
-./Config/run.sh "Config tests"
 
 ./IoTHubManager/run.sh "IoT Hub Manager tests"
 
+./StorageAdapter/run.sh "Storage Adapter Tests"
```
Comment (ppathan) [inline scripts/build]: Does order matter?

File: IoTHubManager/resources/jobs/Method_Chiller_Reboot.json [hunk 33]
```diff
@@ -0,0 +1,9 @@
+﻿{
```
Comment (ppathan) [inline IoTHubManager/resources/jobs/Method_Chiller_Reboot.json]: tabs size is wrong

File: IoTHubManager/Simulation.cs [hunk 34]
```diff
@@ -0,0 +1,107 @@
...
+        {
+            var devices = GetSimulatedDevices();
+            var deviceModels = devices["DeviceModels"];
+ 
+            foreach (JObject device in deviceModels)
+            {
+                if (device["Id"].ToString() == Constants.SimulatedDevices.SIMULATED_DEVICE)
+                {
+                    healthyDeviceNo = device["Count"].ToObject<int>();
+                    //device["Count"] = (healthyDeviceNo + 1).ToString();    Currently, we are using existing
```
Comment (ppathan) [inline IoTHubManager/Simulation.cs]: You might want to check your editor settings since it is not putting space after // in few files in this change

File: IoTHubManager/CreateDeviceTests.cs (hunk 27 above)
Comment (sushilraje) [reply IoTHubManager/CreateDeviceTests.cs]: Yes, it is called for tear down

File: IoTHubManager/Simulation.cs (hunk 34 above)
Comment (sushilraje) [reply IoTHubManager/Simulation.cs]: ok

File: IoTHubManager/Constants.cs (hunk 25 above)
Comment (sushilraje) [reply IoTHubManager/Constants.cs]: Yes, I will see to this

File: IoTHubManager/resources/jobs/Method_Chiller_Reboot.json (hunk 33 above)
Comment (sushilraje) [reply IoTHubManager/resources/jobs/Method_Chiller_Reboot.json]: Fixed

File: scripts/build (hunk 32 above)
Comment (sushilraje) [reply scripts/build]: no, I changed it for debugging my taxis failures.

File: IoTHubManager/Simulation.cs (hunk 34 above)
Comment (sushilraje) [reply IoTHubManager/Simulation.cs]: yes I checked it.

File: IoTHubManager/DevicePropertiesTest.cs (hunk 29 above)
Comment (sushilraje) [reply IoTHubManager/DevicePropertiesTest.cs]: WAIT is used in multiple places. Its 6000 ms. I want to wait for 30 secs. Hence the calculation

File: IoTHubManager/resources/jobs/Tags.json (hunk 31 above)
Comment (sushilraje) [reply IoTHubManager/resources/jobs/Tags.json]: Fixed
